### ISSUE TITLE
Getting an immutable instance doesn't respect the throwForInvalidImmutableState setting in every instantiation case, also @Check methods don't validate required attributes

### ISSUE THREAD

**Issue by matt-isaman** [immutables/immutables#751]
For this class:
```java
package com.foo;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.io.Serializable;

import org.immutables.value.Value.Check;
import org.immutables.value.Value.Immutable;
import org.immutables.value.Value.Parameter;
import org.immutables.value.Value.Style;

@SuppressWarnings("javadoc")
@Immutable
@Style(privateNoargConstructor = true,
       throwForInvalidImmutableState = IllegalArgumentException.class)
public abstract class Foo implements Serializable {

    private static final long serialVersionUID = 1L;

    @JsonProperty("foo")
    @Parameter
    public abstract String foo();

    @Check
    public void validate() throws IllegalArgumentException {
        if ("".equals(foo())) {
            throw new IllegalArgumentException("foo was empty");
        }
    }
}
```
And this test:
```java
package com.foo;


import com.fasterxml.jackson.databind.ObjectMapper;

import org.junit.Test;


@SuppressWarnings("javadoc")
public class ImmutableFooTest {

    @Test(expected = IllegalArgumentException.class)
    public void factoryNull() {
        ImmutableFoo.of(null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void factoryBlank() {
        ImmutableFoo.of("");
    }

    @Test(expected = IllegalArgumentException.class)
    public void builderNull() {
        ImmutableFoo.builder().foo(null).build();
    }

    @Test(expected = IllegalArgumentException.class)
    public void builderBlank() {
        ImmutableFoo.builder().foo("").build();
    }

    @Test(expected = IllegalArgumentException.class)
    public void builderNothingSet() {
        ImmutableFoo.builder().build();
    }

    @Test(expected = IllegalArgumentException.class)
    public void copyOfNull() {
        ImmutableFoo.copyOf(null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void fromNull() {
        ImmutableFoo.builder().from(null).build();
    }

    @Test(expected = IllegalArgumentException.class)
    public void withNull() {
        ImmutableFoo.of("s").withFoo(null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void withBlank() {
        ImmutableFoo.of("s").withFoo("");
    }

    @Test(expected = IllegalArgumentException.class)
    public void jsonEmpty() throws Exception {
        final ObjectMapper mapper = new ObjectMapper();
        final ImmutableFoo foo = mapper.readValue("{}", ImmutableFoo.class);
        foo.validate();
    }

    @Test(expected = IllegalArgumentException.class)
    public void jsonNull() throws Exception {
        final ObjectMapper mapper = new ObjectMapper();
        final ImmutableFoo foo = mapper.readValue("{\"foo\":null}", ImmutableFoo.class);
        foo.validate();
    }

    @Test(expected = IllegalArgumentException.class)
    public void jsonBlank() throws Exception {
        final ObjectMapper mapper = new ObjectMapper();
        final ImmutableFoo foo = mapper.readValue("{\"foo\":\"\"}", ImmutableFoo.class);
        foo.validate();
    }
}
```
The results are:
```
factoryNull() -> FAIL
factoryBlank() -> PASS
builderNull() -> FAIL
builderBlank() -> PASS
builderNothingSet() -> PASS
copyOfNull() -> FAIL
fromNull() -> FAIL
withNull() -> FAIL
withBlank() -> PASS
jsonEmpty() -> FAIL
jsonNull() -> FAIL
jsonBlank() -> PASS
```
All non-json failures are due to:
`java.lang.Exception: Unexpected exception, expected<java.lang.IllegalArgumentException> but was<java.lang.NullPointerException>`
The json failures were due to no exception being thrown:
`java.lang.AssertionError: Expected exception: java.lang.IllegalArgumentException`

This makes complete sense looking at how the `ImmutableFoo` code is generated. But when I use the `throwForInvalidImmutableState` setting I expected the only exception to come out of all the possible ways to get instances to be the exception class I specified. Also, I feel like in the case where `throwForInvalidImmutableState` isn't specified it might be nice to only throw IllegalStateException just so there is a single exception type for being bad at getting an Immutable* instance. The message of that Exception can always indicate it was thrown due to a `null` being passed.

With the way this is currently implemented people using `Immutable*` classes will need to catch `NullPointerException` and `[ IllegalStateException | throwForInvalidImmutableState ]` whenever they create an instance. I think it would be nicer to just need to catch a single "validation failed" exception (one of `[ IllegalStateException | throwForInvalidImmutableState ]`) for any and all issues related to creating an instance.

I copy pasted the generated class and edited it a bit to see what would have to change to get it to throw a single exception for a bad instance. The (edited) diff output for those changes is below.
```
***************
    GENERATED
-------
    MY EDITS
***************
```
```
***************
    private ImmutableFoo(String foo) {
!     this.foo = Objects.requireNonNull(foo, "foo");
    }
-------
    private ImmutableFoo(String foo) {
!     this.foo = foo;
    }
***************
    public final ImmutableFoo withFoo(String value) {
      if (this.foo.equals(value)) return this;
!     String newValue = Objects.requireNonNull(value, "foo");
!     return validate(new ImmutableFoo(this, newValue));
    }
-------
    public final ImmutableFoo withFoo(String value) {
      if (this.foo.equals(value)) return this;
!     return validate(new ImmutableFoo(this, value));
    }
***************
-------
+   @Override
+   public void validate() throws IllegalArgumentException {
+     checkRequiredAttributes(this);
+     super.validate();
+   }
  
    private static ImmutableFoo validate(ImmutableFoo instance) {
      instance.validate();
      return instance;
    }
  
+   private static void checkRequiredAttributes(ImmutableFoo instance) throws IllegalArgumentException {
+     if (Objects.isNull(instance.foo())) {
+       throw new IllegalArgumentException(formatRequiredAttributesMessage(instance));
+     }
+   }
+ 
+   private static String formatRequiredAttributesMessage(ImmutableFoo instance) {
+     List<String> attributes = Lists.newArrayList();
+     if (Objects.isNull(instance.foo())) attributes.add("foo");
+     return "Cannot build Foo, some of required attributes are not set " + attributes;
+   }
***************
      public final Builder from(Foo instance) {
!       Objects.requireNonNull(instance, "instance");
        foo(instance.foo());
        return this;
      }
-------
      public final Builder from(Foo instance) {
!       if(Objects.isNull(instance)) {
!         throw new IllegalArgumentException("Cannot build Foo, from a null instance.");
!       }
        foo(instance.foo());
        return this;
      }
***************
      public final Builder foo(String foo) {
!       this.foo = Objects.requireNonNull(foo, "foo");
!       initBits &= ~INIT_BIT_FOO;
        return this;
      }
-------
      public final Builder foo(String foo) {
!       this.foo = foo;
!       if(Objects.nonNull(foo)) {
!         initBits &= ~INIT_BIT_FOO;
!       }
        return this;
      }
```
With the changes above all the tests passed. There may be a better way to do this where the `checkRequiredAttributes` and `formatRequiredAttributesMessage` methods aren't more or less duplicated in the builder (if present) and the class itself. But this was just a quick check to see how much would have to change.

Thoughts?

**Comment by matt-isaman** [comment]
My request, as I remember it, was motivated by my desire to provide a clean API to my consumers for instantiating Immutable objects. This API would provide a single exception indicating instantiation had failed.

Right now my consumers have two options:
1. Check all required attributes for `null` before passing them through to the builder/factory methods, or
2. Catch NPE _and_ the custom validation exception on every `.build()` call.

I dislike these options and would prefer my consumers only had to catch the validation failed exception. This has the advantage of concentrating all of the checks, for nulls or otherwise, in one place and would allow classes to wrap up instantiation failures of all kinds, including NPEs, in one exception.

I could achieve this now by:
1. Marking all attributes as `@Nullable` and then doing all the required checks in my `@Check` method. But doing it that was seems to slightly defeat the purpose of this library. It is also misleading to the users since not all of the attributes marked with `@Nullable` would _actually_ be nullable. Or,
2. Use the `@Value.Style(validationMethod=ValidationMethod.NONE)` style that you mentioned and do all my checks in the `@Check` method. This at least prevents my needing to inaccurately mark things `@Nullable`.

All that said these two options seemed a bit hacky. I opened this issue to see if there was a way to allow for customizing types to wrap all exceptions during construction into the validation failed exception.

**Comment by elucash** [comment]
There a way to hijack null check to throw the different exception, so, in theory, you can override both NPE and ISE to a single exception. In addition to `throwForInvalidImmutableState` style, there's
the capability rewrite imports, see an example of such extension: https://github.com/immutables/immutables/tree/master/value-fixture/src/org/immutables/fixture/generatorext . Realizing that this not optimal, I just added `throwForNullPointer` style flag to customize the NPE exception, which should be a lot simpler (to be released in next 2.6 milestone or final).
Having that, there's hardly anything we can do to change the lifecycle aspects of objects without me losing sanity.

For the cases where completely custom handling is desired, you can dive into implementing [encodings](http://immutables.github.io/encoding.html) for many different types, this probably lacks the ability to override handling for all types, and this might not be practical.

> Catch NPE and the custom validation exception on every .build() call.

NPEs would be raised when you invoke initializer with null value, before `.build()` is called, build should throw only ISE

```java
try {
// this is not that terrible if people don't care about `null`
} catch (NullPointerException | IllegalStateException buildFailed) {
```
